Bottom: 09af7bfe4db00d19d6b052e13d6dd5aa75b57e7c
Top:    b2fc6f16e8651478a4a3e16f19ac77eaf28d22b0
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-10-20 22:39:58 +0100

Refresh of gl-use-a-vbo-to-avoid-data-cop

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 52d738a..35fa289 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -61,10 +61,28 @@ hidgl_new_triangle_array (void)
 void
 hidgl_init_triangle_array (triangle_buffer *buffer)
 {
-  glEnableClientState (GL_VERTEX_ARRAY);
-  glVertexPointer (3, GL_FLOAT, 0, buffer->triangle_array);
+  GLenum errCode;
+  const GLubyte *errString;
+
   buffer->triangle_count = 0;
   buffer->coord_comp_count = 0;
+
+  glEnableClientState (GL_VERTEX_ARRAY);
+  glGenBuffers (1, &buffer->vbo_name);
+  glBindBuffer (GL_ARRAY_BUFFER, buffer->vbo_name);
+  glBufferData (GL_ARRAY_BUFFER, TRIANGLE_ARRAY_BYTES, NULL, GL_STATIC_DRAW);
+
+  buffer->triangle_array = glMapBuffer (GL_ARRAY_BUFFER, GL_WRITE_ONLY);
+
+  if ((errCode = glGetError()) != GL_NO_ERROR) {
+      errString = gluErrorString(errCode);
+     fprintf (stderr, "OpenGL Error: %s\n", errString);
+  }
+
+  if (buffer->triangle_array == NULL) {
+    printf ("Couldn't map VBO.. sorry, don't know how best to handle this gracefully\n");
+    exit (1);
+  }
 }
 
 void
@@ -73,9 +91,20 @@ hidgl_flush_triangles (triangle_buffer *buffer)
   if (buffer->triangle_count == 0)
     return;
 
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+
+  glEnableClientState (GL_VERTEX_ARRAY);
+  glVertexPointer (2, GL_FLOAT, 0, NULL); // buffer->triangle_array);
   glDrawArrays (GL_TRIANGLES, 0, buffer->triangle_count * 3);
-  buffer->triangle_count = 0;
-  buffer->coord_comp_count = 0;
+  glDisableClientState (GL_VERTEX_ARRAY);
+
+//  buffer->triangle_count = 0;
+//  buffer->coord_comp_count = 0;
+
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer->vbo_name);
+
+  hidgl_init_triangle_array (buffer);
 }
 
 void
diff --git a/src/hid/common/hidgl.h b/src/hid/common/hidgl.h
index a3e966f..8fcf26e 100644
--- a/src/hid/common/hidgl.h
+++ b/src/hid/common/hidgl.h
@@ -24,10 +24,13 @@
 #define __HIDGL_INCLUDED__
 
 #define TRIANGLE_ARRAY_SIZE 5461
+#define TRIANGLE_ARRAY_BYTES (3 * 3 * TRIANGLE_ARRAY_SIZE)
 typedef struct {
-  GLfloat triangle_array [3 * 3 * TRIANGLE_ARRAY_SIZE];
+//  GLfloat triangle_array [3 * 3 * TRIANGLE_ARRAY_SIZE];
+  GLfloat *triangle_array;
   unsigned int triangle_count;
   unsigned int coord_comp_count;
+  GLuint vbo_name;
 } triangle_buffer;
 
 extern triangle_buffer buffer;
diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 278dd4a..f0572cd 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -1161,6 +1161,11 @@ ghid_port_drawing_area_expose_event_cb (GtkWidget * widget,
   hidgl_flush_triangles (&buffer);
   glPopMatrix ();
 
+
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer.vbo_name);
+
   draw_grid ();
 
   hidgl_init_triangle_array (&buffer);
@@ -1181,6 +1186,9 @@ ghid_port_drawing_area_expose_event_cb (GtkWidget * widget,
 
   hidgl_flush_triangles (&buffer);
 
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer.vbo_name);
 
   if (gdk_gl_drawable_is_double_buffered (pGlDrawable))
     gdk_gl_drawable_swap_buffers (pGlDrawable);
diff --git a/src/hid/gtk/gui-pinout-preview.c b/src/hid/gtk/gui-pinout-preview.c
index a086b19..75f4804 100644
--- a/src/hid/gtk/gui-pinout-preview.c
+++ b/src/hid/gtk/gui-pinout-preview.c
@@ -238,6 +238,10 @@ ghid_pinout_preview_expose (GtkWidget * widget, GdkEventExpose * ev)
   hidgl_flush_triangles (&buffer);
   glPopMatrix ();
 
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer.vbo_name);
+
   if (gdk_gl_drawable_is_double_buffered (pGlDrawable))
     gdk_gl_drawable_swap_buffers (pGlDrawable);
   else
